include/rmm/detail/stack_trace.hpp

```diff
@@ -63,9 +65,35 @@ class stack_trace {
     if (strings.get() == nullptr) {
       os << "But no stack trace could be found!" << std::endl;
     } else {
-      ///@todo: support for demangling of C++ symbol names
-      for (int i = 0; i < st.stack_ptrs.size(); ++i) {
-        os << "#" << i << " in " << strings.get()[i] << std::endl;
+      // Iterate over the stack pointers converting to a string
+      for (int i = 0; i < st.stack_ptrs.size(); i++) {
```

Comment:
I think we can make this code more beautiful. Discussed with @codereport . Suggestion is to replace the `for` with a `std::transform` over `st.stack_ptrs()`.

And use an IILE to clean up the internal `if`, something like

```
auto const str = [&] {
    if (dladdr(st.stack_ptrs[i], &info)) {
        int status = -1;
        // Demangle the name. This can occasionally fail
        std::unique_ptr<char, decltype(&::free)> demangled(
            abi::__cxa_demangle(info.dli_sname, nullptr, 0, &status), &::free);
        // If it fails, fallback to the dli_name.        
        if (status == 0 or info.dli_sname) {
            auto const name = status == 0 ? demangled.get() : info.dli_sname;
            return name + " from " << info.dli_fname;
        }
    }
    return "";
} ();
```

Reply:
I implemented the IILE to clean up the string construction but I couldnt come up with a good way to use `std::transform` here that was cleaner than the current implementation. Since we need both the value and the index we would need to use the binary_op overload of `std::transform` with one iterator over the values and something similar to `boost::counting_iterator` for the index.

Maybe you are better with these functions in `<algorithm>` than I am, but I couldn't find an elegant way to create the indexing iterator that didn't require creating a second object like a `vector`. It's a shame because I was hoping to use `std::ostream_iterator` for the output which would have been nice. If I'm missing something here, let me know.